# Case Study 14 
## Chris Irwin
### 04/19/2018

<center>Experimenting with Neural Network Design using Higgs Boson Data</center>

## Introduction
On 4 July 2012, the ATLAS and CMS experiments at CERN's Large Hadron Collider announced they had each observed a new particle. This particle is consistent with the Higgs boson predicted by the Standard Model. According to the CERN or The European Organization of Nuclear Research, the Standard Model explains how the basic building blocks of matter interact. <sup>1</sup> The Higgs boson, as proposed within the Standard Model, is the simplest manifestation of the Brout-Englert-Higgs mechanism. Other types of Higgs bosons are predicted by other theories that go beyond the Standard Model.<sup>2</sup>

In October of 2013 the Nobel prize in physics was awarded jointly to François Englert and Peter Higgs "for the theoretical discovery of a mechanism that contributes to our understanding of the origin of mass of subatomic particles, and which recently was confirmed through the discovery of the predicted fundamental particle, by the ATLAS and CMS experiments at CERN's Large Hadron Collider." <sup>2</sup> This discovery is important because scientists believe that the Higgs boson is the particle that gives all matter its mass.<sup>3</sup>

In the following case study, we will use neural networks and various experiment settings to work with data created from the Higgs Boson experiment. Using the Keras and Tensorflow packages we will show the benefit of using multiple combinations of levels and activations to attempt to find a model to validate the existence of the boson particle. To test the model's overall ability we will be using three different values. The first is an ROC value or Receiver Operationg Characteristic, this calculation is most commonly used as a way to help show the ability and performance of a binary classifier.<sup>9</sup> The second is a loss function, which will be calculated using the Keras package built in loss function of binary_crossentropy. This loss function calculates the probability is the true label, and the vien distribution is the predicted value of the current model.<sup>10</sup> The final value will be an accuracy calculation, this calculation is built into the Keras pacakge and is different from the ROC calculation because the accuracy funcation uses batchs instead of the entire data set to test results. 

## Methods 
### Tensorflow Package
Tensorflow is an open source program that was developed by Google to help with the creation of neural networks.<sup>4</sup> The package the ability to leverage both the computers CPU as well as its GPU in order to speed up the process of training and testing data.

### Keras Package
Keras is a high-level neural networks API, written in Python and capable of running on top of TensorFlow, CNTK, or Theano. It was developed with a focus on enabling fast experimentation.<sup>5</sup>The Keras package has four guiding principles which are user friendliness, modularity, easy extensibility and ability to work with Python. These guiding principles have created an easy to use package that makes creating deep learning neural networks and gives the ability for the data scientists to tune different parameters for enhanced performance.  

## Building Neural Networks
The first step in building neural networks is to add the Keras and Tensorflow packages to you project. The benefits of adding both packages are explained above in greater detail. The next step is to import a number of sub-packages which will help with creating and developing a neural network. 

* Sequential - Allows for individual layers to be added to a neural network.
* Activation - Imports multiple functions that can be passed to create layers.
* Dense - Helps to implement a complex operation of activation and kernels, which are weights matrix created by the layers
* SGD - This package includes the optimizer for Stochastic gradient descent.

In [2]:
import keras
import tensorflow as tf
import numpy as np
import pandas as pd

from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.optimizers import RMSprop
from keras.optimizers import Adagrad
from keras.optimizers import Nadam
from sklearn.metrics import roc_auc_score


/Users/cmirwin/anaconda2/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using Theano backend.


After adding the packages to the case study, the next step is import data into a pandas data frame and then reshape the data to the expected format. The reshape of the data is a very important step as the keras and tensorflow packages are extremely particular when it comes the shape of the data.

In [3]:
data=pd.read_csv("./HIGGS.csv",nrows=5000000,header=None)
test_data=pd.read_csv("./HIGGS.csv",nrows=200000,header=None,skiprows=6000000)

In [4]:
y=np.array(data.loc[:,0])
x=np.array(data.loc[:,1:])
x_test=test_data.loc[:,1:]
y_test=test_data.loc[:,0]

This case study will take a look at number of scenarios, combinations and import totals to attempt to increase the model’s ability to correctly predict outcomes. The first combination we will try is a neural net with 5 layers and using an initializer of Uniform and the sigmoid activation function on all layers.

In [48]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)



After setting up the layers of the neural network the next step is to test and train the data to see what the possible outcomes are. Additionally, the fit function that is included with the Keras packages uses Epochs. According to the Keras documentation an Epoch is an arbitrary cutoff, generally defined as "one pass over the entire dataset", used to separate training into distinct phases, which is useful for logging and periodic evaluation. <sup>6</sup> Also we use batch sizes of 5,000 to test the model's overall accuracy. 

In [49]:
model.fit(x, y, epochs=5, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/5
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6915 - acc: 0.5297
Epoch 2/5
5000000/5000000 [==============================] - 60s 12us/step - loss: 0.6915 - acc: 0.5298
Epoch 3/5
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 4/5
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 5/5
200000/200000 [==============================] - 2s 8us/step
0.5610751371472434
[0.6909302428364754, 0.5333549976348877]


Overall the neural net above is set up with four layers using a batch size of 5,000 and 5 epochs, the neural net created a model with an overall accuracy of 53.3%, with an ROC score of .5610 and a loss value of .6909 when compared to the test set, which was created earlier in the analysis. This is a very poor performing model, to improve our score the next step is to alter the layers and attempt to alter the outcome. Additionally, In the following neural net, we will increase the number of Epochs to see if there is any additional information gained from the additional runs.

In [50]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 2/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 3/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 4/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 5/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 6/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 7/15
5000000/5000000 [==============================] - 61s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 8/15
5000000/5000000 [==============================] - 62s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 9/15
5000000/5000000 [==============================] - 62s 12us/step - loss: 0.6914 - acc: 0.5298
Epoch 10/15
5000000/5000000 [==========================

After looking at the results we see that adding Epochs did not really increase the overall accuracy of the model, but we do see an increase in the ROC score, with this additional increase we will continue to use this higher number of Epochs going forward in order to see what additional information we can glean. The next step we will try to improve the score with is to remove a single layer and attempt to see if that increases the overall model accuracy.

In [51]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6923 - acc: 0.5261
Epoch 2/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6917 - acc: 0.5289
Epoch 3/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6915 - acc: 0.5297
Epoch 4/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6855 - acc: 0.5477
Epoch 5/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6480 - acc: 0.6182
Epoch 6/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6400 - acc: 0.6325
Epoch 7/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6351 - acc: 0.6420
Epoch 8/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6249 - acc: 0.6553
Epoch 9/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.6123 - acc: 0.6677
Epoch 10/15
5000000/5000000 [==========================

After removing a single the layer with a node size of fifty we see a considerable increase in the model’s overall accuracy, which is an increase of sixteen percentage points, .192 increase in ROC score and a .12 decrease in the loss value. With this in mind the next model we will run we will decrease the node size per layer from 200 to 150 to see what the overall effect is to the model.

In [52]:
model = Sequential()
model.add(Dense(150, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('sigmoid'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6918 - acc: 0.5291
Epoch 2/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6916 - acc: 0.5295
Epoch 3/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6916 - acc: 0.5296
Epoch 4/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6914 - acc: 0.5297
Epoch 5/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6876 - acc: 0.5419
Epoch 6/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6498 - acc: 0.6157
Epoch 7/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6397 - acc: 0.6328
Epoch 8/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6327 - acc: 0.6452
Epoch 9/15
5000000/5000000 [==============================] - 42s 8us/step - loss: 0.6209 - acc: 0.6592
Epoch 10/15
5000000/5000000 [==============================] - 4

Looking at the results from the model which was run above we find that by decreasing the node size of each layer creates a similar overall accuracy, ROC score and loss value when applied to the test data set. The interesting part in this section of the analysis is that similar results were achieved using a lower node count, but the smaller nodes sizes ran in a considerably less amount of time by averaging approximately 18-20 seconds less per run. 

## Alter Activation Methods

The next step in this case study will be to examine the usage of different activations. Using the model above that had the highest accuracy, which was the model with 3 layers of 200 per node, we will apply multiple activations in an attempt to raise the score. Keras gives access to a number of different activations and we will focus specifically on the relu, tanh and linear activations. Additionally, the other two models are subjected to the same process of altering the activations with code and results available in the appendix located at the end of this case study.

### relu Activation

In [53]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 30s 6us/step - loss: 0.6164 - acc: 0.6492
Epoch 2/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5550 - acc: 0.7116
Epoch 3/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5297 - acc: 0.7307
Epoch 4/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5137 - acc: 0.7414
Epoch 5/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5022 - acc: 0.7489
Epoch 6/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4949 - acc: 0.7537
Epoch 7/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4896 - acc: 0.7572
Epoch 8/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4861 - acc: 0.7596
Epoch 9/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4829 - acc: 0.7615
Epoch 10/15
5000000/5000000 [==============================] - 3

The results above are very compelling, by removing the sigmoid activation and replacing it with relu we found there to be a large increase in the ROC score, a seven percentage point increase in accuracy and a .10 decrease in the loss value. An additional benefit to the relu function also seems to be an overall decrease in the amount of run time per Epoch of almost 30 seconds per run.

Continuing the attempt to improve the score of the model we will next compare the results of the model when we change the relu activation for the tanh activation to see if there is any increase in accruacy or decrease in run time. 

### tanh

In [54]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('tanh'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.6375 - acc: 0.6344
Epoch 2/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5968 - acc: 0.6791
Epoch 3/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5785 - acc: 0.6933
Epoch 4/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5654 - acc: 0.7032
Epoch 5/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.5552 - acc: 0.7110
Epoch 6/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5477 - acc: 0.7167
Epoch 7/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5414 - acc: 0.7216
Epoch 8/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.5356 - acc: 0.7259
Epoch 9/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.5304 - acc: 0.7297
Epoch 10/15
5000000/5000000 [==========================

The results of the tanh are still an overall improvement over the use of sigmoid activations with an overall increase in the ROC Score, an increase of 14 percentage points in accuracy and decrease in the loss value of .7. When compared to the relu result the tanh there is a glaring increase in the run time with each run lasting approximently 55 seconds on average, with an overall decrease in accuracy and overall increase in loss. 

Next we will look at the linear activation to see how it performs and if there is any increase in overall model percision. 

### linear

In [55]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('linear'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
#rms = RMSprop(lr=0.001, rho=0.9, epsilon=None, decay=1e-6)
#ada = Adagrad(lr=0.01, epsilon=None, decay=0.0)
#nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 28s 6us/step - loss: 0.6445 - acc: 0.6268
Epoch 2/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6385 - acc: 0.6395
Epoch 3/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6384 - acc: 0.6400
Epoch 4/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6383 - acc: 0.6401
Epoch 5/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6383 - acc: 0.6404
Epoch 6/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6382 - acc: 0.6403
Epoch 7/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6382 - acc: 0.6404
Epoch 8/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6382 - acc: 0.6405
Epoch 9/15
5000000/5000000 [==============================] - 29s 6us/step - loss: 0.6382 - acc: 0.6405
Epoch 10/15
5000000/5000000 [==============================] - 2

The final activation that we attempted had results that were the lowest amongst the three activations we tried. Even when compared to the original model with sigmoid activation we find that the model’s accuracy decreased by approximately five percentage points and gained approximately .07 in loss value. Interestingly the run times per Epoch improved considerably and were the best out of all the activations used. 

## Altering Batch Sizes

After finding that the relu activation is gives us the highest accuracy the next item to see if we can increase the overall accuracy is to increase and decrease the total batch size. We will run two different scenarios with batch sizes of 25,000 records a batch of 75. 

### Batch of 25,000

In [56]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=25000)
score = model.evaluate(x_test, y_test, batch_size=25000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.6831 - acc: 0.5513
Epoch 2/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.6275 - acc: 0.6451
Epoch 3/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.6003 - acc: 0.6753
Epoch 4/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5848 - acc: 0.6882
Epoch 5/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.5722 - acc: 0.6980
Epoch 6/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.5637 - acc: 0.7047
Epoch 7/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5568 - acc: 0.7103
Epoch 8/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5495 - acc: 0.7160
Epoch 9/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.5427 - acc: 0.7210
Epoch 10/15
5000000/5000000 [==============================] - 3

### Batches of 75

In [57]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=75)
score = model.evaluate(x_test, y_test, batch_size=75)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 98s 20us/step - loss: 0.5594 - acc: 0.7098
Epoch 2/15
5000000/5000000 [==============================] - 107s 21us/step - loss: 0.5299 - acc: 0.7336
Epoch 3/15
5000000/5000000 [==============================] - 111s 22us/step - loss: 0.5192 - acc: 0.7410
Epoch 4/15
5000000/5000000 [==============================] - 114s 23us/step - loss: 0.5128 - acc: 0.7454
Epoch 5/15
5000000/5000000 [==============================] - 116s 23us/step - loss: 0.5082 - acc: 0.7480
Epoch 6/15
5000000/5000000 [==============================] - 115s 23us/step - loss: 0.5047 - acc: 0.7504
Epoch 7/15
5000000/5000000 [==============================] - 116s 23us/step - loss: 0.5017 - acc: 0.7523
Epoch 8/15
5000000/5000000 [==============================] - 115s 23us/step - loss: 0.4990 - acc: 0.7540
Epoch 9/15
5000000/5000000 [==============================] - 116s 23us/step - loss: 0.4968 - acc: 0.7554
Epoch 10/15
5000000/5000000 [==================

After running two different batch sizes we can see that by increasing the batch size to 25,000 records per batch decreased the model’s accuracy by approximately 2 percentage points and increased the loss value by approximately .03. Additionally, decreasing the batch size to 75 slightly decreased the accuracy by .6 percentage point and slightly increased the loss value. Overall the performance was very similar between the batches of 75 vs 5,000, but the run time between the two is dramatically different with the batches of 75 taking on average 2 minutes per Epoch or an increase of 150%.

## Altering Kernal Initializers

As we explained earlier in this case study there are a number of different variables that can be altered to increase the accuracy and effectiveness of the model. These values are available because of the different type data that can be ran through the neural net process, whether the data is image related, scientific results or sports statistics different data requires different algorithms to get the best results. According to the Keras documentation the Initializations define the way to set the initial random weights of Keras layers. <sup>7</sup>

In the following part of the case study we will run the previous defined model with 3 layers of node sizes 200 with batch sizes of 5,000 with three different initializer values. Previously in this case study we have used a constant kernel initializer of uniform and have achieved very good overall results. Now we will run the same model and compare the results by changing the kernel initializer to use either orthogonal, VarianceScaling or truncated_normal initializer values.

### Orthogonal

In [58]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='orthogonal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='orthogonal'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='orthogonal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='orthogonal')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 30s 6us/step - loss: 0.5923 - acc: 0.6776
Epoch 2/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5391 - acc: 0.7241
Epoch 3/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5143 - acc: 0.7411
Epoch 4/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5013 - acc: 0.7496
Epoch 5/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4943 - acc: 0.7539
Epoch 6/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4896 - acc: 0.7571
Epoch 7/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4861 - acc: 0.7595
Epoch 8/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4834 - acc: 0.7608
Epoch 9/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4811 - acc: 0.7626
Epoch 10/15
5000000/5000000 [==============================] - 3

After updating the kernel initializer value to Orthogonal, we see the results are very similar to the when we ran the process with the Uniform kernel initializer value. All four categories, ROC Score, accuracy, loss value and time per Epoch show little difference between the two initializer values.

### VarianceScaling

In [59]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='VarianceScaling')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='VarianceScaling'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='VarianceScaling'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='VarianceScaling')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 30s 6us/step - loss: 0.5942 - acc: 0.6760
Epoch 2/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5437 - acc: 0.7205
Epoch 3/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5207 - acc: 0.7371
Epoch 4/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5067 - acc: 0.7462
Epoch 5/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4980 - acc: 0.7518
Epoch 6/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4926 - acc: 0.7552
Epoch 7/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4889 - acc: 0.7577
Epoch 8/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4858 - acc: 0.7596
Epoch 9/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4834 - acc: 0.7611
Epoch 10/15
5000000/5000000 [==============================] - 3

The next step of the analysis is to change Orthogonal to an initialization value of VarianceScaling. Once again, we see a very minimal amount of difference between the results of both the Uniform, Orthogonal and VarianceScaling. Overall we see a slight descrease in ROC Score, a descrease in accuracy and a slight increase in loss value. 


### Truncated_Normal

In [60]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 38s 8us/step - loss: 0.5956 - acc: 0.6726
Epoch 2/15
5000000/5000000 [==============================] - 38s 8us/step - loss: 0.5373 - acc: 0.7251
Epoch 3/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5129 - acc: 0.7416
Epoch 4/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5000 - acc: 0.7502
Epoch 5/15
5000000/5000000 [==============================] - 33s 7us/step - loss: 0.4928 - acc: 0.7550
Epoch 6/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4877 - acc: 0.7582
Epoch 7/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4844 - acc: 0.7603
Epoch 8/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4815 - acc: 0.7622
Epoch 9/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4791 - acc: 0.7637
Epoch 10/15
5000000/5000000 [==============================] - 3

The final initializer value we attempted is the truncated_normal. This value gave us our first lowering of the loss value by approximately .05 when compared to all three of the previously used initializer values. Additionally, we saw an increase in the model’s accuracy with an approximately 1 percentage point and an increase in ROC Score. Moving forward to the next step of our case study we will use an initialization value of truncated_normal


## Altering Neural Net Optimizers 
Finally, after altering layer sizes and counts, changing activation types, testing different batch sizes and running multiple kernels we are going to attempt to improve the accuracy of the model with trying different optimizers. The choice of optimizer is an incredibly important one as it will commonly affect the overall performance of your model. Below we will attempt to build the model using three specific RMSProp, Nadam and Adagrad

*RMSProp - RMSprop is an unpublished, adaptive learning rate method proposed by Geoff Hinton in Lecture 6e of his Coursera Class. <sup>8</sup>

*Nadam – Is similar to RMSProp with an adjustment for momentum by using the Nesterov accelerated gradient <sup>8</sup>

*Adagrad - is an algorithm for gradient-based optimization that does just this: It adapts the learning rate to the parameters, performing larger updates for infrequent and smaller updates for frequent parameters. <sup>8</sup>

### RMSProp

In [61]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('relu'))

rms = RMSprop(lr=0.001, rho=0.9, epsilon=None, decay=1e-6)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=rms)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.6273 - acc: 0.6448
Epoch 2/15
5000000/5000000 [==============================] - 34s 7us/step - loss: 0.5710 - acc: 0.7039
Epoch 3/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5503 - acc: 0.7210
Epoch 4/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5381 - acc: 0.7303
Epoch 5/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5314 - acc: 0.7358
Epoch 6/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5274 - acc: 0.7395
Epoch 7/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5243 - acc: 0.7421
Epoch 8/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5217 - acc: 0.7443
Epoch 9/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5216 - acc: 0.7456
Epoch 10/15
5000000/5000000 [==============================] - 3

Once we altered the optimizer to be RMSProp we found that there was a decrease in ROC Score of approximately .02, a decrease in accuracy of approximately .02, an increase in loss function of approximately .05 and finally an increase in the amount of time it took for each Epoch to run. 

Next we tested the optimizer of Adagrad to see what effect it has on the models performance.
### Adagrad

In [62]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

ada = Adagrad(lr=0.01, epsilon=None, decay=0.0)
#nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=ada)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5903 - acc: 0.6784
Epoch 2/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5489 - acc: 0.7168
Epoch 3/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5331 - acc: 0.7292
Epoch 4/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5236 - acc: 0.7359
Epoch 5/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5169 - acc: 0.7404
Epoch 6/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5120 - acc: 0.7435
Epoch 7/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5078 - acc: 0.7460
Epoch 8/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5045 - acc: 0.7481
Epoch 9/15
5000000/5000000 [==============================] - 35s 7us/step - loss: 0.5016 - acc: 0.7499
Epoch 10/15
5000000/5000000 [==============================] - 3

After altering the optimizer to Adagrad and rerunning the experiment we saw a slight increase in ROC score and accuracy from the new optimizer when compared to the RMSProp run. In comparing the new optimizer to our current best test we found that there is still a decrease in ROC Score and accuracy with an increase in loss. 

Finally, we will run the experiment one last time to see if the Nadam optimizer has any positive effect on the overall model.

### Nadam

In [63]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=nad)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5768 - acc: 0.6912
Epoch 2/15
5000000/5000000 [==============================] - 38s 8us/step - loss: 0.5211 - acc: 0.7369
Epoch 3/15
5000000/5000000 [==============================] - 36s 7us/step - loss: 0.5008 - acc: 0.7503
Epoch 4/15
5000000/5000000 [==============================] - 36s 7us/step - loss: 0.4892 - acc: 0.7574
Epoch 5/15
5000000/5000000 [==============================] - 36s 7us/step - loss: 0.4824 - acc: 0.7619
Epoch 6/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.4775 - acc: 0.7650
Epoch 7/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.4740 - acc: 0.7674
Epoch 8/15
5000000/5000000 [==============================] - 36s 7us/step - loss: 0.4707 - acc: 0.7695
Epoch 9/15
5000000/5000000 [==============================] - 36s 7us/step - loss: 0.4685 - acc: 0.7707
Epoch 10/15
5000000/5000000 [==============================] - 3

The Nadam optimizer produced the best results of any model we have produced so far. Additionally, it performed the best of all three of the optimizers with an overall ROC score of .859 and an accuracy of .775. When compared to the highest achieved accuracy in this case study the Nadam model increased the accuracy by approximately .05.

After numerous attempts at maximizing the score of our model we have found that having the following settings has produced a very successful model.

* Record Count – 5 million
* Batch Size – 5000
* Layers – 3
* Node Size - 200
* Epochs - 15
* Kernel Initializer – truncated_normal
* Activation – Relu
* Optimizer – Nadam

Earlier in the case study we attempted to lower the node size to see if it would help with improving the model. The results of that experiment were a slight overall change in the model’s accuracy and ROC score. In a final attempt to maximize the score we attempted to increase the size node size to see if the additional data enhanced the ability of the model. Below we will attempt two additional analysis, the first is increasing the node size to 350 and the second is increasing the node size to 750. 


### Nadam with Nodes of 350

In [5]:
model = Sequential()
model.add(Dense(350, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(350, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(350, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=nad)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 89s 18us/step - loss: 0.5721 - acc: 0.6948
Epoch 2/15
5000000/5000000 [==============================] - 88s 18us/step - loss: 0.5137 - acc: 0.7419
Epoch 3/15
5000000/5000000 [==============================] - 92s 18us/step - loss: 0.4928 - acc: 0.7554
Epoch 4/15
5000000/5000000 [==============================] - 95s 19us/step - loss: 0.4818 - acc: 0.7625
Epoch 5/15
5000000/5000000 [==============================] - 97s 19us/step - loss: 0.4749 - acc: 0.7669
Epoch 6/15
5000000/5000000 [==============================] - 97s 19us/step - loss: 0.4701 - acc: 0.7700
Epoch 7/15
5000000/5000000 [==============================] - 98s 20us/step - loss: 0.4662 - acc: 0.7725
Epoch 8/15
5000000/5000000 [==============================] - 98s 20us/step - loss: 0.4629 - acc: 0.7747
Epoch 9/15
5000000/5000000 [==============================] - 99s 20us/step - loss: 0.4601 - acc: 0.7765
Epoch 10/15
5000000/5000000 [==========================

Upon the completion of the model with node sizes of 350 we found there to an increase in accuracy, by approximently .02, and increase in ROC Score, by approximently .2, and slight decrease in the loss value of .003. By increasing the node size by 150, or 75% over the previous model we saw an increase in time per Epoch approximently 60 seconds or 200% on average. With the increase model’s overall accuracy measures we continued to increase the node size to 750 to see if the model would continue to increase.

### Increasing Node Size to 750

In [7]:
model = Sequential()
model.add(Dense(750, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(750, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(750, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=nad)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 313s 63us/step - loss: 0.5693 - acc: 0.6981
Epoch 2/15
5000000/5000000 [==============================] - 313s 63us/step - loss: 0.5057 - acc: 0.7475
Epoch 3/15
5000000/5000000 [==============================] - 326s 65us/step - loss: 0.4849 - acc: 0.7609
Epoch 4/15
5000000/5000000 [==============================] - 323s 65us/step - loss: 0.4741 - acc: 0.7679
Epoch 5/15
5000000/5000000 [==============================] - 325s 65us/step - loss: 0.4666 - acc: 0.7725
Epoch 6/15
5000000/5000000 [==============================] - 327s 65us/step - loss: 0.4607 - acc: 0.7764
Epoch 7/15
5000000/5000000 [==============================] - 325s 65us/step - loss: 0.4560 - acc: 0.7792
Epoch 8/15
5000000/5000000 [==============================] - 320s 64us/step - loss: 0.4510 - acc: 0.7823
Epoch 9/15
5000000/5000000 [==============================] - 321s 64us/step - loss: 0.4468 - acc: 0.7848
Epoch 10/15
5000000/5000000 [=================

Upon the completion of the final model using node sizes of 750 we find there to be a decrease in the models accuracy and ROC score. Additionally, the loss value was slightly higher, with the largest difference being the amount of time per Epoch. The average Epoch lasted 5.5 minutes, which is approximately 300% increase in run time when compared to the model with node sizes of 350. With the increase in time and resources, but a decrease in all the scoring attributes we chose to stop experiment and end with the model that gave us the highest results.


## Conclusion

The case study above used a number of different variables to attempt to create a neural network based on the Higgs Boson data set. One of the first variables we looked at was the creation of a model with different layers, our original model had five-layers with our second attempt having only four. The decrease in layers created a significant gain in both accuracy the ROC scores. For example, the second model with three layers had a ROC score of .7654 and the original five-layer model had a ROC score of .5730 respectively, which an approximate increase of .192. Additionally, the accuracy of the four-layer model is .6966 and the original five-layer model is .5333 respectively, which is an increase of approximately .1633. Through multiple iterations we found that by adding an extra layer to the model and using the original settings for initialization kernel and optimizer there was added noise, which the final result was a less accurate model with a larger loss function. To further test whether the four-layer model was the most efficient we tested a three-layer model, the code and results are available in the appendix, using the original settings with node sizes of 200 and the final settings with node sizes of 200 and 350. Overall the three-layer model performed better with the original settings, but when starting to alter additional features the model ended up performing slightly worse. With this in mind we chose to move forward in the case with using a 3-layer neural net with node sizes of 200. 

After deciding on a 3-layer model with node sizes of 200, we then set out to try multiple activation values. These three separate values were relu, tanh and linear. The results of this experiment were to use the relu activation, which resulted in a ROC score of .8528 and an accuracy of approximately .7688. These values were a dramatic increase from the original sigmoid activation that was previously attempted with an increase of approximately .09 for the model’s ROC score and a .07 increase in accuracy of the model. 

Next, we continued the testing of different variables by changing the initialization kernel that was used for each of the layers. For this case study we used the uniform, truncated_normal, VarianceScaling and Orthogonal initialization kernels. After multiple iterations we found that the truncated_normal initialization kernel gave us increased results over the other three values with a ROC score of .8544, which is an increase of approximately .002 and an approximately .0014 increase in the model’s accuracy when compared to a relu model with four-layers and a uniform initializer.  

Finally, we tested multiple optimizer values, which where the Stochastic Gradient Descent or SGD,  RMSProp, Adagrad and Nadam. When compared to the model’s using SGD both the RMSProp and Adagrad models performed worse and created ROC and accuracy scores that were slightly less than the SGD model. The Nadam model produced the best model with a final ROC curve of .8544 and an accuracy score of .7702 respectively.

In conclusion this case study took an extensive look at neural networks and how different settings can affect a model’s overall performance. After multiple iterations we found that a four-layer model with an activation of relu for the first three layers and sigmoid for the final layer, all with node sizes of 350 with initialization kernels of truncated_normal and a Nadam optimizer allowed us to find a final model with a ROC score of .8616 an accuracy score of .7772 and a loss value of .4704. Overall we found that the largest increase in the model’s accuracy came from the testing of different layers, with the second most being attributed to the changes in activation. 


## References 
1. https://home.cern/about/physics/standard-model
2. https://home.cern/topics/higgs-boson
3. https://www.cnn.com/2011/12/13/world/europe/higgs-boson-q-and-a/index.html
4. https://en.wikipedia.org/wiki/TensorFlow
5. https://keras.io/
6. https://keras.io/getting-started/faq/#what-does-sample-batch-epoch-mean
7. https://keras.io/initializers/
8. http://ruder.io/optimizing-gradient-descent/index.html
9. http://www.dataschool.io/roc-curves-and-auc-explained
10. https://en.wikipedia.org/wiki/Cross-entropy

## Appendix

### Model 1 with relu

In [64]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.6576 - acc: 0.5882
Epoch 2/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5685 - acc: 0.7013
Epoch 3/15
5000000/5000000 [==============================] - 32s 6us/step - loss: 0.5368 - acc: 0.7250
Epoch 4/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5150 - acc: 0.7402
Epoch 5/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.5011 - acc: 0.7494
Epoch 6/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4930 - acc: 0.7547
Epoch 7/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4875 - acc: 0.7582
Epoch 8/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4838 - acc: 0.7606
Epoch 9/15
5000000/5000000 [==============================] - 31s 6us/step - loss: 0.4805 - acc: 0.7627
Epoch 10/15
5000000/5000000 [==============================] - 3

### Model 1 with tanh

In [65]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('tanh'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.6456 - acc: 0.6201
Epoch 2/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5984 - acc: 0.6783
Epoch 3/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5804 - acc: 0.6928
Epoch 4/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5689 - acc: 0.7015
Epoch 5/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5591 - acc: 0.7089
Epoch 6/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5504 - acc: 0.7154
Epoch 7/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5435 - acc: 0.7201
Epoch 8/15
5000000/5000000 [==============================] - 54s 11us/step - loss: 0.5376 - acc: 0.7243
Epoch 9/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5328 - acc: 0.7277
Epoch 10/15
5000000/5000000 [==========================

### Model 1 with linear

In [66]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('linear'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(50, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6498 - acc: 0.6160
Epoch 2/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6386 - acc: 0.6393
Epoch 3/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6384 - acc: 0.6399
Epoch 4/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6383 - acc: 0.6401
Epoch 5/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6383 - acc: 0.6403
Epoch 6/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6382 - acc: 0.6404
Epoch 7/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6382 - acc: 0.6404
Epoch 8/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6382 - acc: 0.6405
Epoch 9/15
5000000/5000000 [==============================] - 27s 5us/step - loss: 0.6382 - acc: 0.6406
Epoch 10/15
5000000/5000000 [==============================] - 2

### Model 3 with relu

In [67]:
model = Sequential()
model.add(Dense(150, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.6179 - acc: 0.6465
Epoch 2/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.5560 - acc: 0.7107
Epoch 3/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.5312 - acc: 0.7288
Epoch 4/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.5139 - acc: 0.7409
Epoch 5/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.5024 - acc: 0.7487
Epoch 6/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.4954 - acc: 0.7534
Epoch 7/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.4908 - acc: 0.7563
Epoch 8/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.4871 - acc: 0.7589
Epoch 9/15
5000000/5000000 [==============================] - 20s 4us/step - loss: 0.4845 - acc: 0.7603
Epoch 10/15
5000000/5000000 [==============================] - 2

### Model 3 with tanh

In [68]:
model = Sequential()
model.add(Dense(150, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('tanh'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('tanh'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.6352 - acc: 0.6354
Epoch 2/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5951 - acc: 0.6815
Epoch 3/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5796 - acc: 0.6932
Epoch 4/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5685 - acc: 0.7011
Epoch 5/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5596 - acc: 0.7078
Epoch 6/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5528 - acc: 0.7131
Epoch 7/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5459 - acc: 0.7182
Epoch 8/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5393 - acc: 0.7233
Epoch 9/15
5000000/5000000 [==============================] - 37s 7us/step - loss: 0.5343 - acc: 0.7270
Epoch 10/15
5000000/5000000 [==============================] - 3

### Model 3 with linear

In [69]:
model = Sequential()
model.add(Dense(150, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('linear'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(150, kernel_initializer='uniform'))
model.add(Activation('linear'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6458 - acc: 0.6240
Epoch 2/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6385 - acc: 0.6395
Epoch 3/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6384 - acc: 0.6399
Epoch 4/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6383 - acc: 0.6402
Epoch 5/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6383 - acc: 0.6402
Epoch 6/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6382 - acc: 0.6404
Epoch 7/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6382 - acc: 0.6405
Epoch 8/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6382 - acc: 0.6406
Epoch 9/15
5000000/5000000 [==============================] - 18s 4us/step - loss: 0.6382 - acc: 0.6406
Epoch 10/15
5000000/5000000 [==============================] - 1

## Model with 3-layers and Node size of 200

In [9]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='uniform')) # X_train.shape[1] == 15 here
model.add(Activation('sigmoid'))
model.add(Dense(200, kernel_initializer='uniform'))
model.add(Activation('sigmoid'))
model.add(Dense(1, kernel_initializer='uniform')) 
model.add(Activation('sigmoid'))


sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=sgd)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 38s 8us/step - loss: 0.6817 - acc: 0.5520
Epoch 2/15
5000000/5000000 [==============================] - 41s 8us/step - loss: 0.6445 - acc: 0.6238
Epoch 3/15
5000000/5000000 [==============================] - 39s 8us/step - loss: 0.6383 - acc: 0.6365
Epoch 4/15
5000000/5000000 [==============================] - 39s 8us/step - loss: 0.6304 - acc: 0.6492
Epoch 5/15
5000000/5000000 [==============================] - 39s 8us/step - loss: 0.6168 - acc: 0.6636
Epoch 6/15
5000000/5000000 [==============================] - 40s 8us/step - loss: 0.6027 - acc: 0.6764
Epoch 7/15
5000000/5000000 [==============================] - 40s 8us/step - loss: 0.5948 - acc: 0.6827
Epoch 8/15
5000000/5000000 [==============================] - 40s 8us/step - loss: 0.5898 - acc: 0.6867
Epoch 9/15
5000000/5000000 [==============================] - 41s 8us/step - loss: 0.5856 - acc: 0.6898
Epoch 10/15
5000000/5000000 [==============================] - 3

## 3-layer model with Node Size 200 and final settings

In [10]:
model = Sequential()
model.add(Dense(200, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(200, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=nad)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 22s 4us/step - loss: 0.5878 - acc: 0.6825
Epoch 2/15
5000000/5000000 [==============================] - 24s 5us/step - loss: 0.5420 - acc: 0.7229
Epoch 3/15
5000000/5000000 [==============================] - 26s 5us/step - loss: 0.5227 - acc: 0.7366
Epoch 4/15
5000000/5000000 [==============================] - 25s 5us/step - loss: 0.5098 - acc: 0.7452
Epoch 5/15
5000000/5000000 [==============================] - 22s 4us/step - loss: 0.5012 - acc: 0.7505
Epoch 6/15
5000000/5000000 [==============================] - 22s 4us/step - loss: 0.4950 - acc: 0.7543
Epoch 7/15
5000000/5000000 [==============================] - 22s 4us/step - loss: 0.4902 - acc: 0.7576
Epoch 8/15
5000000/5000000 [==============================] - 24s 5us/step - loss: 0.4866 - acc: 0.7597
Epoch 9/15
5000000/5000000 [==============================] - 24s 5us/step - loss: 0.4841 - acc: 0.7613
Epoch 10/15
5000000/5000000 [==============================] - 2

## 3-layer model with Node Size 350 and final settings

In [11]:
model = Sequential()
model.add(Dense(350, input_dim=x.shape[1], kernel_initializer='truncated_normal')) # X_train.shape[1] == 15 here
model.add(Activation('relu'))
model.add(Dense(350, kernel_initializer='truncated_normal'))
model.add(Activation('relu'))
model.add(Dense(1, kernel_initializer='truncated_normal')) 
model.add(Activation('sigmoid'))

nad = Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
model.compile(loss='binary_crossentropy', metrics=['accuracy'], optimizer=nad)

model.fit(x, y, epochs=15, batch_size=5000)
score = model.evaluate(x_test, y_test, batch_size=5000)
print(roc_auc_score(y_test,model.predict(x_test)))
print(score)

Epoch 1/15
5000000/5000000 [==============================] - 54s 11us/step - loss: 0.5791 - acc: 0.6901
Epoch 2/15
5000000/5000000 [==============================] - 57s 11us/step - loss: 0.5276 - acc: 0.7334
Epoch 3/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.5077 - acc: 0.7468
Epoch 4/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.4966 - acc: 0.7539
Epoch 5/15
5000000/5000000 [==============================] - 55s 11us/step - loss: 0.4892 - acc: 0.7585
Epoch 6/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.4842 - acc: 0.7614
Epoch 7/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.4803 - acc: 0.7642
Epoch 8/15
5000000/5000000 [==============================] - 56s 11us/step - loss: 0.4774 - acc: 0.7659
Epoch 9/15
5000000/5000000 [==============================] - 58s 12us/step - loss: 0.4749 - acc: 0.7676
Epoch 10/15
5000000/5000000 [==========================